thor/src/channel.rs

```diff
@@ -0,0 +1,623 @@
+pub mod protocols;
+
+pub use protocols::create::{BuildFundingPsbt, SignFundingPsbt};
+use protocols::{close, create, punish::punish, splice, update};
+
+use crate::{
+    keys::{OwnershipKeyPair, OwnershipPublicKey},
+    signature, step, step_wallet,
+    transaction::{ptlc, FundingTransaction},
+    Balance, BroadcastSignedTransaction, ChannelState, GetRawTransaction, MedianTime, Message,
+    NewAddress, Ptlc, PtlcPoint, PtlcSecret, ReceiveMessage, RevokedState, Role, SendMessage,
+    Splice, SplitOutput, StandardChannelState,
+};
+use ::serde::{Deserialize, Serialize};
+use anyhow::{anyhow, bail, Result};
+use bitcoin::{Address, Amount, Transaction, Txid};
+use futures::{
+    future::{Either, FutureExt},
+    pin_mut, Future,
+};
+use genawaiter::sync::Gen;
+use std::{convert::TryInto, time::Duration};
+
+#[cfg(test)]
+mod tests;
+
+#[macro_export]
+macro_rules! update {
+    ($transport:expr, $state:expr) => {{
+        let transport = $transport;
+        let state = $state;
+
+        let (transport, state) = step!(transport, state);
+        let (transport, state) = step!(transport, state);
+        let (_, updated_channel) = step!(transport, state);
+
+        updated_channel
+    }};
+}
```

Comment:
:upside_down_face: I liked it better when this was local to the update function. That way you can ensure that it always appears before it is used too.

Reply:
And you won't need the `#[macro_export]` statement.

Reply:
Yeah, I did also. I am unsure why this stopped building as it was. Will have another look.

Reply:
Interesting. When `update()` was in lib.rs the macro definition could be at the end of the function but once that code was in `channel.rs` it had to be at the top of the function. So rustc must parse sub-modules differently to `lib.rs`? Do you know what why the code worked when it was in rustc?

Reply:
I encountered a similar problem, but I just fixed it the same way you did by declaring it at the top of the function where it is used. In `lib.rs` it did work at the end of the function, but only if it was annotated with `#[macro_export]` I think.